rspace/src/main/scala/coop/rchain/rspace/history/TrieCache.scala

```diff
@@ -0,0 +1,181 @@
+package coop.rchain.rspace.history
+import coop.rchain.rspace.Blake2b256Hash
+
+import scala.collection.immutable.{Map => ImmutableMap}
+import scala.collection.immutable.Seq
+import scala.collection.mutable.Map
+
+private[rspace] trait CachedItem[TItem] {
+  def asOption: Option[TItem]
+}
+
+private[rspace] case class LoadedItem[TItem](value: TItem) extends CachedItem[TItem] {
+  override def asOption: Option[TItem] = Some(value)
+}
+
+private[rspace] case class StoredItem[TItem](value: TItem) extends CachedItem[TItem] {
+  override def asOption: Option[TItem] = Some(value)
+}
+
+private[rspace] case class AbsentItem[TItem]() extends CachedItem[TItem] {
+  override def asOption: Option[TItem] = None
+}
+
+private[rspace] class TrieCache[T, K, V](
+    trieStore: ITrieStore[T, K, V],
+    val trieBranch: Branch,
+    trieStoreOwner: Boolean = false
+) extends ITrieStore[T, K, V] {
```

Comment:
have you had the chance to measure the memory impact on rnode?

Reply:
To minimize memory pressure I'm using WeakReferences, so gc will be able to collect all additional instances if needed. On our loop with 3000 iterations:

With TrieCache peak memory usage: 0.48gb
Old version peak memory usage: 0.46gb

Profiler snapshots comparison shows that additional memory related to new classes takes less than 10 megabytes, however most of alive objects aren`t related to rspace, so it's hard to measure memory impact with precision of kilobytes.